tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala

```diff
@@ -25,9 +27,114 @@ import at.forsyte.apalache.tla.typecomp.subbuilder.{
  *     not be scope-correct, since x would have to be typed as Int within the scope defined by this \A operator. Thus,
  *     such an expression cannot be constructed by the builder.
  *
+ * Note that the guarantees are void if `useTrustedEx` is used.
+ *
  * @author
  *   Jure Kukovec
  */
 class ScopedBuilder
-    extends BaseBuilder with BoolBuilder with ArithmeticBuilder with SetBuilder with SeqBuilder
-    with LiteralAndNameBuilder
+    extends BaseBuilder with BoolBuilder with ArithmeticBuilder with SetBuilder with SeqBuilder with ActionBuilder
+    with LiteralAndNameBuilder {
+
+  /*
+  To use the builder outside of testing scenarios, where the expressions aren't written from scratch,
+  it is necessary to be able to mark certain expressions as "trusted", e.g. when transforming and recombining invariants,
+  or parts of Init/Next.
+  While `build` is safe as a unidirectional implicit conversion from TBuilderInstruction to TLaEx,
+  the inverse, `useTrustedEx`, needs to be explicit, to stress the fact that it should be used rarely,
+  typically at most once per transformation, on the initial input.
+   */
+  /**
+   * Creates a `TBuilderInstruction` from a precomputed `TlaEx`.
+   *
+   * Should be used sparingly, and only for expressions that have already passed static analysis.
+   *
+   * Voids correctness guarantees.
+   */
+  def useTrustedEx(ex: TlaEx): TBuilderInstruction = ex.point[TBuilderInternalState]
+
+  /** x' = y */
+  def primeEq(x: TBuilderInstruction, y: TBuilderInstruction): TBuilderInstruction = eql(prime(x), y)
+
+  type TypedParam = (OperParam, TlaType1)
+
+  /** Parameters to TLA operators can be typed as OperT1 exactly at the top-level, and only if arity > 0 */
+  private def isAcceptableParamType(canContainOper: Boolean): TlaType1 => Boolean = {
+    case FunT1(arg, res)         => isAcceptableParamType(false)(arg) && isAcceptableParamType(false)(res)
+    case SetT1(elem)             => isAcceptableParamType(false)(elem)
```

Comment:
Can we get a functional description as Scaladoc?

The current comment is more explanatory about the implementation, and should probably go under `case OperT1(args, res)` the function body?